grpc/server/src/connection_initiator.h

```diff
@@ -0,0 +1,82 @@
+#include <future>
+
+#include "connection_result.h"
+#include "log.h"
+
+namespace dronecore {
+namespace backend {
+
+template <typename DroneCore>
+class ConnectionInitiator
+{
+public:
+    ConnectionInitiator() {}
+    ~ConnectionInitiator() {}
+
+    bool start(DroneCore &dc, const int port)
+    {
+        init_mutex();
+        init_timeout_logging(dc);
+
+        _discovery_future = wrapped_register_on_discover(dc);
+
+        if (!add_udp_connection(dc, port)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    void wait()
+    {
+        _discovery_future.wait();
+    }
+
+private:
+    void init_mutex()
+    {
+        _discovery_promise = std::make_shared<std::promise<uint64_t>>();
+    }
+
+    void init_timeout_logging(DroneCore &dc) const
+    {
+        dc.register_on_timeout([](uint64_t uuid) {
+            LogInfo() << "Device timed out [UUID: " << uuid << "]";
+        });
+    }
+
+    bool add_udp_connection(DroneCore &dc, const int port)
+    {
+        dronecore::ConnectionResult connection_result = dc.add_udp_connection(port);
+
+        if (connection_result != ConnectionResult::SUCCESS) {
+            LogErr() << "Connection failed: " << connection_result_str(connection_result);
+            return false;
+        }
+
+        return true;
+    }
+
+    std::future<uint64_t> wrapped_register_on_discover(DroneCore &dc)
+    {
+        auto future = _discovery_promise->get_future();
+
+        LogInfo() << "Waiting to discover device...";
+
+        dc.register_on_discover([this](uint64_t uuid) {
+            std::call_once(_discovery_flag, [this, uuid]() {
```

Comment:
Does this mean that it is only called once? So it is not called anymore after a timeout and then discovery.

Reply:
I think it doesn't matter here, as I only want to initiate the connection? And the `dronecore` plugin (or telemetry? I don't remember) is the one sending the state to the client.

Reply:
And actually as soon as the first connection is established, the `dronecore` plugin will probably overwrite the `register_on_discover` callback since, for now, there is only one callback saved at a time.